### PR TITLE
Support non-tensor inputs and outputs for checkpointed functions.

### PR DESCRIPTION
**Pull request by pritamdamania87**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#52422 Support non-tensor inputs and outputs for checkpointed functions.**

As mentioned in https://github.com/pytorch/pytorch/issues/52415,
`torch.utils.checkpoint` doesn't support checkpointing for functions which have
non-tensor inputs and outputs.

This PR resolves this issue by ensuring the autograd machinery ignores the
non-tensor inputs and outputs and processes the tensors accordingly.

In addition to this, I've also updated `gradcheck` to handle non-Tensor outputs appropriately.

Differential Revision: [D26507228](https://our.internmc.facebook.com/intern/diff/D26507228/)

#Closes: https://github.com/pytorch/pytorch/issues/52415

### CODE REVIEW

File: torch/utils/checkpoint.py
Comment (soulitzer) [inline torch/utils/checkpoint.py:73]: Would the tensor inputs still be freed on the backward pass here when `retain_graph=False`?
Maybe we could pick out the tensor arguments from inputs and still use `save_for_backward` in those cases and only use ctx to store non-tensor inputs

File: torch/utils/checkpoint.py
Comment (soulitzer) [inline torch/utils/checkpoint.py:118]: Is there a reason we originally returned `inp` for grad if `inp` is not a Tensor? Why does it need to be `None` now? Just curious

**Comment by gchanan** [comment]
I'll let @albanD comment but off the top of my head this doesn't look like the right approach.  The issue has to do with checkpoints, but we are now pushing all the complexity into the autograd engine.  Is there some approach we can take where we construct/wrap the function in a smarter way?

**Comment by gchanan** [comment]
sorry, I should say "custom function implementation" rather than "autograd engine"

**Comment by pritamdamania87** [comment]
> I'll let @albanD comment but off the top of my head there might be a nicer approach. The issue has to do with checkpoints, but we are now pushing all the complexity into the custom function. Is there some approach we can take where we construct/wrap the function in a smarter way?

I guess the PR and the issue mention checkpointing, but I thought of fixing this in general for custom functions as well. Pipeline parallelism uses a bunch of custom functions and it would be nice if this was supported in custom functions as well instead of just checkpointing.

**Comment by gchanan** [comment]
I see -- could you point to use cases outside of checkpointing?

**Comment by pritamdamania87** [comment]
> I see -- could you point to use cases outside of checkpointing?

This is the use case in pipeline parallelism: https://github.com/pytorch/pytorch/blob/master/torch/distributed/pipeline/sync/checkpoint.py#L249. It uses its own custom autograd function for checkpointing since it reduces GPU idle time. Full explanation can be found at the top of the checkpoint.py file.

File: test/test_utils.py
Comment (soulitzer) [inline test/test_utils.py:302]: nit: Rest -> Reset

**Comment by soulitzer** [comment]
Maybe add a test for when none of the inputs/outputs are tensors? You probably also want to fix the clang-tidy failure, but LGTM otherwise. 

I'll let @albanD give the final approval since he mentioned he wanted to take a quick look once he's back.

File: torch/csrc/autograd/custom_function.cpp
Comment (albanD) [inline torch/csrc/autograd/custom_function.cpp:108]: What if the user returned an undefined Tensor on which he wants history to be tracked?

File: torch/csrc/autograd/custom_function.cpp
Comment (albanD) [inline torch/csrc/autograd/custom_function.cpp:149]: Why do you need to check if var is defined here? You bail out at the beginning of the loop in the other case no?

File: torch/csrc/autograd/python_function.cpp
Comment (albanD) [inline torch/csrc/autograd/python_function.cpp:370]: Can't this be confused with a user provided undefined Tensor?

File: test/test_utils.py
Comment (albanD) [inline test/test_utils.py:306]: nit: you can do all of that in a single backward call

File: torch/utils/checkpoint.py
Comment (albanD) [inline torch/utils/checkpoint.py:107]: you can use enumerate to remove the need for tensors_idx

File: torch/utils/checkpoint.py
Comment (albanD) [inline torch/utils/checkpoint.py:132]: nit: tensor_args_index

**Review by albanD — COMMENTED** [review]
If we go with this, tests for both the python and c++ version of the custom Function are missing as well as doc update for both custom Functions and checkpoint.

File: torch/csrc/autograd/custom_function.cpp
Comment (malfet) [inline torch/csrc/autograd/custom_function.cpp:122]: Both num_tensors and output_nr are `unit32_t`, please remove the cast
```suggestion
      AT_ASSERT(num_tensors == output_nr);
```

File: test/test_utils.py
Comment (malfet) [inline test/test_utils.py:302]: To repeat @soulitzer's suggestion
```suggestion
        # Reset grads, run without checkpoint and validate we receive same grads.
```

**Comment by pritamdamania87** [comment]
> Maybe add a test for when none of the inputs/outputs are tensors?

Is this a valid use case from a user point of view though? If there are no tensors to a module, there is no need to run backward through the module and I'm not sure if a custom autograd function makes sense? For example checkpointing today doesn't allow this: https://github.com/pytorch/pytorch/blob/master/torch/utils/checkpoint.py#L24

**Review by pritamdamania87 — COMMENTED** [review]
> If we go with this, tests for both the python and c++ version of the custom Function are missing as well as doc update for both custom Functions and checkpoint.

I made these changes in the latest update to the PR for python custom functions. I didn't intend to make these changes for the cpp custom functions as well since those actually still have strict requirements for outputs being Variables. I was wondering if the scope can only be python custom function support? The changes I made in custom_function.cpp were only to ensure python custom functions can support a variety of datatypes.

File: torch/csrc/autograd/custom_function.cpp
Comment (albanD) [reply torch/csrc/autograd/custom_function.cpp:108]: The thing is that this code is in common for both python and c++
But I could see users trying to track gradients for their custom objects and using undefined Tensors as a "cheap" object to give to the autograd. Not that I would recommend doing that :p 

I think that if you only want to make the python version work for checkpoint, it will be simpler for sure to do a wrapper on checkpoint.

Otherwise, I think we will have to update this function to take list of optional Tensor to be able to properly mark the outputs that should not be considered and/or use mark_non_differentiable. But that might have more perf implications...

Comment (pritamdamania87) [reply torch/csrc/autograd/custom_function.cpp:108]: > But I could see users trying to track gradients for their custom objects and using undefined Tensors as a "cheap" object to give to the autograd. Not that I would recommend doing that :p

Hmm I'm still not sure how an undefined Tensor can be used to track autograd history? If the output of a function is an undefined tensor, can't it be ignored for autograd purposes (as we do here)? 

> I think that if you only want to make the python version work for checkpoint, it will be simpler for sure to do a wrapper on checkpoint.

I guess the goal here is to support this for python custom functions in general and not just checkpoint. In particular classes like these: https://github.com/pytorch/pytorch/blob/master/torch/distributed/pipeline/sync/checkpoint.py#L249.

Comment (albanD) [reply torch/csrc/autograd/custom_function.cpp:108]: My question is that if we only want to support that for python function, is there a way to make this change python only? That will make things easier to review and less likely to break the c++ Function in a subtle way.

**Comment by soulitzer** [comment]
> > Maybe add a test for when none of the inputs/outputs are tensors?
> 
> Is this a valid use case from a user point of view though? If there are no tensors to a module, there is no need to run backward through the module and I'm not sure if a custom autograd function makes sense? For example checkpointing today doesn't allow this: https://github.com/pytorch/pytorch/blob/master/torch/utils/checkpoint.py#L24

Probably not. If we don't expect users to do this, we could add a check similar to the one you linked, but for custom functions in general.

**Comment by pritamdamania87** [comment]
@albanD I updated the PR to get rid of any changes to C++ custom functions and kept the changes python only.

@soulitzer Added tests for none of the inputs/outputs being tensors. Looks like checkpoint supports this and only raises a warning.

File: torch/csrc/autograd/python_function.cpp
Comment (albanD) [inline torch/csrc/autograd/python_function.cpp:388]: nit: this could be moved inside the if statement below

File: torch/csrc/autograd/python_function.cpp
Comment (albanD) [inline torch/csrc/autograd/python_function.cpp:572]: you can re-used obj here.

File: torch/autograd/function.py
Comment (albanD) [inline torch/autograd/function.py:176]: Why can it be used to saved inputs?
The extending doc explicitly says that you should use save_for_backward for inputs: https://pytorch.org/docs/stable/notes/extending.html

**Review by albanD — COMMENTED** [review]
Looks much better thanks!

I don't see any mention of what the backward() function will get as input when the forward returns non-Tensor objects.
Are they going to be None gradients? (just like we do for the grad_input)

**Review by pritamdamania87 — COMMENTED** [review]
> I don't see any mention of what the backward() function will get as input when the forward returns non-Tensor objects.
Are they going to be None gradients? (just like we do for the grad_input)

Updated the docs to include this as well.

File: torch/autograd/function.py
Comment (albanD) [reply torch/autograd/function.py:176]: I think this sentence means that you can do `ctx.foo =` to save anything that is not input/output.
But this is most likely not clear indeed.

File: torch/autograd/function.py
Comment (albanD) [inline torch/autograd/function.py:189]: If we just ignore them, then what about having more constraint here that you must provide all Tensor outputs first, then as many non-Tensor objects that will be ignored?
That way the indexing of the Tensors properly match.

Beyond that, I am not 100% convinced about just ignoring these. It will make the forward/backward signature non-symetric which I am not sure I like..
cc @gchanan do you have an opinion on this?

Comment (pritamdamania87) [reply torch/autograd/function.py:189]: > If we just ignore them, then what about having more constraint here that you must provide all Tensor outputs first, then as many non-Tensor objects that will be ignored?
That way the indexing of the Tensors properly match.

I'm not sure if we need to worry about Tensor indexing. From an autograd point of view it only sees that the autograd function outputs tensors and no other values. 

> Beyond that, I am not 100% convinced about just ignoring these. It will make the forward/backward signature non-symetric which I am not sure I like..

I don't think the forward and backward signatures were symmetric earlier either right? You can write a forward function which takes as input 5 tensors and outputs only 3 tensors. In that case, your backward function would get as input 3 tensors, but it would return 5 tensors (for grads of the inputs).

**Comment by soulitzer** [comment]
Can we add a test for gradcheck? I don't think gradcheck currently expects outputs to be non tensors

File: torch/autograd/function.py
Comment (pritamdamania87) [reply torch/autograd/function.py:189]: @albanD @gchanan Wondering if you had any more thoughts regarding this specific issue? Would like to get this change in soon to unblock some of our work on pipeline parallelism to support custom types :)

Comment (albanD) [reply torch/autograd/function.py:189]: I think we definitely want to keep this matching between the number of outputs of the forward and the number of inputs in the backward.

File: torch/autograd/function.py
Comment (albanD) [reply torch/autograd/function.py:176]: I am still not sure that this update makes sense. "tensors/inputs" is not really accurate because you can save pretty much everything you want, including intermediary result and now non-tensor outputs as well.

File: torch/autograd/gradcheck.py
Comment (albanD) [inline torch/autograd/gradcheck.py:414]: Why change this? The filtering above should be already removing all the non-Tensor outputs?

File: torch/autograd/gradcheck.py
Comment (albanD) [inline torch/autograd/gradcheck.py:292]: Why do you need to change gradcheck here?
I feel like this should be in a separate PR because that may have some unexpected impact as well....

File: torch/csrc/autograd/python_function.cpp
Comment (albanD) [inline torch/csrc/autograd/python_function.cpp:390]: This should be in the else statement no?

File: torch/csrc/autograd/python_function.cpp
Comment (albanD) [inline torch/csrc/autograd/python_function.cpp:569]: Is it ok for the jit to not trace some values? Won't that create wrong jit IR?

File: torch/utils/checkpoint.py
Comment (albanD) [inline torch/utils/checkpoint.py:100]: The original object is a tuple no? not a list

File: torch/utils/checkpoint.py
Comment (albanD) [inline torch/utils/checkpoint.py:120]: This function is just used here no? Maybe just update the function to take a list as input instead of a tuple?

File: torch/utils/checkpoint.py
Comment (albanD) [inline torch/utils/checkpoint.py:134]: This indexing feels bad. Following the update to include None in the grad_output, that will simplify this code a lot!

File: torch/autograd/function.py
Comment (albanD) [reply torch/autograd/function.py:189]: And having all of them contiguous at the beginning doesn't feel necessary indeed, we can drop this idea.

**Review by albanD — COMMENTED** [review]
I updated my answer below for handling of the non-Tensor outputs.
Also added a few comments inline where the code is a bit surprising.

I haven't looked at the test yet, will do once the grad_output update is in.

File: torch/autograd/gradcheck.py
Comment (pritamdamania87) [reply torch/autograd/gradcheck.py:292]: I added gradcheck as per @soulitzer's suggestion. I guess it makes sense to fix `gradcheck` as well as part of this PR since without this, someone could call `gradcheck` on the output of a custom autograd function and it would fail.

File: torch/csrc/autograd/python_function.cpp
Comment (pritamdamania87) [inline torch/csrc/autograd/python_function.cpp:582]: @suo I was wondering if you had some suggestions on how to test JIT is doing the right thing here :)

File: torch/autograd/function.py
Comment (pritamdamania87) [reply torch/autograd/function.py]: I updated the PR to make this symmetric, but had to touch `custom_function.cpp` for this since that controls how the autograd function metadata is setup. Although, I did it in a way such that it only impacts python custom functions and not cpp custom functions.

File: torch/utils/checkpoint.py
Comment (albanD) [reply torch/utils/checkpoint.py:100]: Ho sorry I though this was coming from the saved_tensors

File: torch/autograd/gradcheck.py
Comment (albanD) [reply torch/autograd/gradcheck.py:292]: Right but is this exercised / tested?

File: torch/autograd/gradcheck.py
Comment (albanD) [reply torch/autograd/gradcheck.py:414]: I am not sure about these changes.
Can you add some tests to make sure it does the right thing if you keep these?

File: torch/autograd/gradcheck.py
Comment (pritamdamania87) [reply torch/autograd/gradcheck.py:292]: Yes, I've added a test for this. See last line of `test_custom_function_non_tensor_inputs_outputs`.

File: torch/csrc/autograd/python_function.cpp
Comment (suo) [reply torch/csrc/autograd/python_function.cpp:582]: I don't have any special insight except for to do some trace checking tests on representative functions (similar to the ones in `test_jit.py`)

File: torch/csrc/autograd/python_function.cpp
Comment (albanD) [inline torch/csrc/autograd/python_function.cpp:392]: double check that the capital version of this macro is not leaking memory

File: torch/csrc/autograd/custom_function.cpp
Comment (albanD) [inline torch/csrc/autograd/custom_function.cpp:13]: nit is_empty ?

**Review by albanD — COMMENTED** [review]
- Change to custom function is good. just nit for naming
- checkpoint doc should be updated to tell what is accepted as input/output and what won't work (nested structures)
- gradcheck change should be reverted as we don't want to support non-tensor output there
- Add a wrapper function to be able to test the new feature with gradcheck

File: torch/utils/checkpoint.py
Comment (albanD) [inline torch/utils/checkpoint.py:165]: nit would -> will

**Review by albanD — APPROVED** [review]
Checked the points from the comments above.

LGTM

**Pull request closed without merging**
